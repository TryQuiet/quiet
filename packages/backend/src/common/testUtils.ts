import fs from 'fs'
import getPort from 'get-port'
import type { Libp2p } from 'libp2p'
import { HttpsProxyAgent } from 'https-proxy-agent'
import path from 'path'
import PeerId from 'peer-id'
import tmp from 'tmp'
import { Config } from '../constants'
import { ConnectionsManager } from '../libp2p/connectionsManager'
import { createCertificatesTestHelper } from '../libp2p/tests/client-server'
import { PermsData } from '@quiet/state-manager'
import WebsocketsOverTor from '../libp2p/websocketOverTor'
import { Tor } from '../torManager'
import {
  createLibp2pAddress,
  createLibp2pListenAddress,
  getPorts,
  Ports,
  torBinForPlatform,
  torDirForPlatform
} from './utils'
import crypto from 'crypto'
import logger from '../logger'
const log = logger('test')

export const rootPermsData: PermsData = {
  certificate:
    'MIIBNjCB3AIBATAKBggqhkjOPQQDAjASMRAwDgYDVQQDEwdaYmF5IENBMCYYEzIwMjEwNjIyMDkzMDEwLjAyNVoYDzIwMzAwMTMxMjMwMDAwWjASMRAwDgYDVQQDEwdaYmF5IENBMFkwEwYHKoZIzj0CAQYIKoZIzj0DAQcDQgAEV5a3Czy+L7IfVX0FpJtSF5mi0GWGrtPqv5+CFSDPrHXijsxWdPTobR1wk8uCLP4sAgUbs/bIleCxQy41kSSyOaMgMB4wDwYDVR0TBAgwBgEB/wIBAzALBgNVHQ8EBAMCAAYwCgYIKoZIzj0EAwIDSQAwRgIhAPOzksuipKyBALt/o8O/XwsrVSzfSHXdAR4dOWThQ1lbAiEAmKqjhsmf50kxWX0ekhbAeCTjcRApXhjnslmJkIFGF2o=+lmBImw3BMNjA0FTlK5iRmVC+w/T6M04Es+yiYL608vOhx2slnoyAwHjAPBgNVHRMECDAGAQH/AgEDMAsGA1UdDwQEAwIABjAKBggqhkjOPQQDAgNIADBFAiEA+0kIz0ny/PLVERTcL0+KCpsztyA6Zuwzj05VW5NMdx0CICgdzf0lg0/2Ksl1AjSPYsy2w+Hn09PGlBnD7TiExBpx',
  privKey:
    'MIGTAgEAMBMGByqGSM49AgEGCCqGSM49AwEHBHkwdwIBAQQgTvNuJL0blaYq6zmFS53WmmOfHshlqn+8wNHDzo4df5WgCgYIKoZIzj0DAQehRANCAARXlrcLPL4vsh9VfQWkm1IXmaLQZYau0+q/n4IVIM+sdeKOzFZ09OhtHXCTy4Is/iwCBRuz9siV4LFDLjWRJLI5+lmBImw3BMNjA0FTlK5iRmVC+w/T6M04Es+yiYL608vOhx2sln'
}

tmp.setGracefulCleanup()

export const testBootstrapMultiaddrs = [
  createLibp2pAddress('abcd.onion', 'QmfLUJcDSLVYnNqSPSRK4mKG8MGw51m9K2v59k3yq1C8s4')
]

export const spawnTorProcess = async (quietDirPath: string, ports?: Ports): Promise<Tor> => {
  const _ports = ports || (await getPorts())
  const torPath = torBinForPlatform()
  const libPath = torDirForPlatform()
  const tor = new Tor({
    appDataPath: quietDirPath,
    torPath: torPath,
    httpTunnelPort: _ports.httpTunnelPort,
    options: {
      env: {
        LD_LIBRARY_PATH: libPath,
        HOME: quietDirPath
      },
      detached: true
    }
  })
  return tor
}

export const createLibp2p = async (peerId: PeerId): Promise<Libp2p> => {
  const pems = await createCertificatesTestHelper('address1.onion', 'address2.onion')

  const port = await getPort()

  return ConnectionsManager.createBootstrapNode({
    peerId,
    listenAddresses: [createLibp2pListenAddress('localhost')],
    bootstrapMultiaddrsList: testBootstrapMultiaddrs,
    agent: new HttpsProxyAgent({ port: 1234, host: 'localhost' }),
    localAddress: createLibp2pAddress('localhost', peerId.toB58String()),
    transportClass: WebsocketsOverTor,
    cert: pems.userCert,
    key: pems.userKey,
    ca: [pems.ca],
    targetPort: port
  })
}

export const createTmpDir = (): tmp.DirResult => {
  return tmp.dirSync({ mode: 0o750, prefix: 'quietTestTmp_', unsafeCleanup: true })
}

export const tmpQuietDirPath = (name: string): string => {
  return path.join(name, Config.QUIET_DIR)
}

export function createFile(filePath: string, size: number) {
  const stream = fs.createWriteStream(filePath)
  const maxChunkSize = 1048576 // 1MB
  stream.on('open', () => {
    if (size < maxChunkSize) {
      stream.write(crypto.randomBytes(size))
    } else {
      const chunks = Math.floor(size / maxChunkSize)
      for (let i = 0; i < chunks; i++) {
        if (size < maxChunkSize) {
          stream.write(crypto.randomBytes(size))
        } else {
          stream.write(crypto.randomBytes(maxChunkSize))
        }
        size -= maxChunkSize
      }
    }
    stream.end()
  })
}
