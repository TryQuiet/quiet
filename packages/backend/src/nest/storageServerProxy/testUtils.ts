export const prepareResponse = (responseData: Partial<Response>) => {
  const ok = responseData.status ? responseData.status >= 200 && responseData.status < 300 : false
  const response: Response = {
    headers: new Headers(),
    ok,
    redirected: false,
    status: 200,
    statusText: '',
    type: 'basic',
    url: '',
    clone: function (): Response {
      throw new Error('Function not implemented.')
    },
    body: null,
    bodyUsed: false,
    arrayBuffer: function (): Promise<ArrayBuffer> {
      throw new Error('Function not implemented.')
    },
    blob: function (): Promise<Blob> {
      throw new Error('Function not implemented.')
    },
    formData: function (): Promise<FormData> {
      throw new Error('Function not implemented.')
    },
    json: function (): Promise<any> {
      throw new Error('Function not implemented.')
    },
    text: function (): Promise<string> {
      throw new Error('Function not implemented.')
    },
    ...responseData,
  }
  return response
}
