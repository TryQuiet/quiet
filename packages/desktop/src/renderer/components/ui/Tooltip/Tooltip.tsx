import React, { ReactElement, ReactFragment } from 'react'
import { styled, Theme } from '@mui/material/styles'
import classNames from 'classnames'

import MuiTooltip, { TooltipProps } from '@mui/material/Tooltip'

// Styling mui tooltip requires workaround: https://mui.com/material-ui/guides/interoperability/#portals
const StyledTooltip = styled(({ className, ...props }: TooltipProps) => (
  <MuiTooltip {...props} classes={{ popper: className }} />
), {})(({ theme }) => ({
  '& .MuiTooltip-tooltip': {
    marginBottom: 5,
    background: theme.palette.colors.trueBlack,
    color: theme.typography.body1.color,
    paddingTop: 12,
    paddingBottom: 12,
    paddingLeft: 16,
    paddingRight: 16,
    borderRadius: 8,
    fontSize: 12,
    fontWeight: 500,

    '&:first-letter': {
      textTransform: 'capitalize',
    }
  },
  '& .MuiTooltip-arrow': {
    '&:before': {
      border: `1px solid ${theme.palette.colors.trueBlack}`
    },
    color: theme.palette.colors.trueBlack
  }
}))

interface CustomTooltipProps {
  children: ReactElement
  title?: string
  titleHTML?: ReactFragment
  interactive?: boolean
  className?: string
  placement?: 'bottom' | 'top' | 'bottom-start' | 'bottom-end' | 'top-start' | 'top-end'
  onClick?: (e: React.MouseEvent) => void
}

export const Tooltip: React.FC<React.ComponentProps<typeof MuiTooltip> & CustomTooltipProps> = ({
  children,
  title,
  titleHTML,
  interactive = false,
  className = '',
  placement = 'bottom',
  onClick = () => {},
  ...props
}) => {
  return (
    <span onClick={e => onClick(e)}>
      <StyledTooltip
        {...props}
        className={className}
        title={
          title.length === 0 ? (
            ''
          ) : (
            <React.Fragment>
              {titleHTML || (
                <span>{title}</span>
              )}
            </React.Fragment>
          )
        }
        placement={placement}
        arrow
        >
        {children}
      </StyledTooltip>
    </span>
  )
}

export default Tooltip
