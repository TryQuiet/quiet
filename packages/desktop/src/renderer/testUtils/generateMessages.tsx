import { DisplayableMessage } from '@quiet/state-manager'

interface IGenerateMessages {
  amount?: number
  type?: number
  message?: string
  nickname?: string
}

const defaults = { amount: 1, type: 1, message: 'message', nickname: 'gringo' }

export const generateMessages = (options: IGenerateMessages = defaults) => {
  let { amount, type, message, nickname } = { ...options }
  amount = amount || defaults.amount
  type = type || defaults.type
  message = message || defaults.message
  nickname = nickname || defaults.nickname
  const messages: DisplayableMessage[] = []
  for (let i = 0; i < amount; i++) {
    messages.push({
      id: `${i}`,
      type,
      message: `${message}${i}`,
      createdAt: 0,
      date: 'string',
      nickname,
    })
  }

  return messages
}
