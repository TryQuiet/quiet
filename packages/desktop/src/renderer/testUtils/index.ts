export * from './generateMessages'
export * from './prepareStore'
export * from './renderComponent'
export * from './socket'

export const sleep = async (timeMs = 1000) => {
  await new Promise<void>(resolve =>
    setTimeout(() => {
      resolve()
    }, timeMs)
  )
}
