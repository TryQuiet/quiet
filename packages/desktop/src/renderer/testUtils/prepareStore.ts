import {
  communities,
  identity,
  users,
  errors,
  messages,
  publicChannels,
  connection,
  settings,
  files,
  StoreKeys as StateManagerStoreKeys,
  Store,
  network,
} from '@quiet/state-manager'
import { StoreKeys } from '../store/store.keys'
import { combineReducers, createStore, applyMiddleware } from 'redux'
import createSagaMiddleware, { Saga, SagaMonitor as SagaMonitorType, Task } from 'redux-saga'
import thunk from 'redux-thunk'
import rootSaga from '../sagas/index.saga'
import { reducer as appReducer } from '../store/handlers/app'
import { socketActions, socketReducer } from '../sagas/socket/socket.slice'
import { modalsReducer } from '../sagas/modals/modals.slice'
import { navigationReducer } from '../store/navigation/navigation.slice'
import MockedSocket from 'socket.io-mock'
import { fork, delay, call, put } from 'typed-redux-saga'

export const reducers = {
  [StateManagerStoreKeys.Communities]: communities.reducer,
  [StateManagerStoreKeys.Identity]: identity.reducer,
  [StateManagerStoreKeys.Users]: users.reducer,
  [StateManagerStoreKeys.Errors]: errors.reducer,
  [StateManagerStoreKeys.Messages]: messages.reducer,
  [StateManagerStoreKeys.PublicChannels]: publicChannels.reducer,
  [StateManagerStoreKeys.Connection]: connection.reducer,
  [StateManagerStoreKeys.Settings]: settings.reducer,
  [StateManagerStoreKeys.Files]: files.reducer,
  [StateManagerStoreKeys.Network]: network.reducer,
  [StoreKeys.App]: appReducer,
  [StoreKeys.Socket]: socketReducer,
  [StoreKeys.Modals]: modalsReducer,
  [StoreKeys.Navigation]: navigationReducer,
}

interface Options {
  effectId: number
  parentEffectId: number
  label?: string
  effect: any
  result: any
}

export interface PrepareStore {
  store: Store
  runSaga: <S extends Saga<any[]>>(saga: S, ...args: Parameters<S>) => Task
  sagaMonitor: SagaMonitor
}

class SagaMonitor {
  effectsTriggeredArray
  effectsResolvedArray
  constructor() {
    this.effectsTriggeredArray = new Map<number, Options>()
    this.effectsResolvedArray = new Map<number, Options>()
  }

  effectTriggered: SagaMonitorType['effectTriggered'] = (options: Options) => {
    this.effectsTriggeredArray.set(options.effectId, options)
  }

  effectResolved: SagaMonitorType['effectResolved'] = (effectId: number, result) => {
    const triggeredEffect = this.effectsTriggeredArray.get(effectId)
    if (!triggeredEffect) return
    this.effectsResolvedArray.set(effectId, { ...triggeredEffect, result })
  }

  public isEffectResolved = (effectName: string) => {
    const parentEffect = Array.from(this.effectsResolvedArray).filter(effect => {
      return effect[1].result?.meta?.name === effectName
    })
    const childrenEffects = Array.from(this.effectsResolvedArray).filter(effect => {
      return effect[1].parentEffectId === parentEffect[0][0]
    })
    return childrenEffects.filter(effect => {
      return effect[1].result === '@@redux-saga/TERMINATE'
    }).length
  }
}

export const prepareStore = async (
  mockedState?: { [key in StoreKeys | StateManagerStoreKeys]?: any },
  mockedSocket?: MockedSocket
): Promise<PrepareStore> => {
  const combinedReducers = combineReducers(reducers)

  const sagaMonitor = new SagaMonitor()
  const sagaMiddleware = createSagaMiddleware({
    sagaMonitor,
  })

  const store = createStore(combinedReducers, mockedState, applyMiddleware(...[sagaMiddleware, thunk]))
  // Fork State manager's sagas (require mocked socket.io-client)
  if (mockedSocket) {
    sagaMiddleware.run(rootSaga)
    // Mock socket connected event
    await sagaMiddleware.run(mockSocketConnectionSaga, mockedSocket).toPromise()
  }

  return {
    store,
    runSaga: sagaMiddleware.run,
    sagaMonitor,
  }
}

function* mockSocketConnectionSaga(socket: MockedSocket): Generator {
  yield* fork(function* (): Generator {
    yield* delay(1000)
    yield* call(() => {
      socket.socketClient.emit('connect')
    })
  })
  yield* put(socketActions.startConnection({ dataPort: 4677, socketIOToken: 'testToken' }))
}
