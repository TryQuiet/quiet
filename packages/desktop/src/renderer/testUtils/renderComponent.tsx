import React, { FC, ReactElement } from 'react'
import { ThemeProvider, StyledEngineProvider } from '@mui/material/styles'
import { Provider } from 'react-redux'

import { render } from '@testing-library/react'

import theme from '../theme'
import store from '../store'
import { Store } from 'redux'
import { DndProvider } from 'react-dnd'
import { HTML5Backend } from 'react-dnd-html5-backend'

interface Props {
  children?: React.ReactNode
}

export const renderComponent = (ui: ReactElement, storeState: Store = store): ReturnType<typeof render> => {
  const Wrapper: FC<Props> = ({ children }) => (
    <StyledEngineProvider injectFirst>
      <ThemeProvider theme={theme}>
        <DndProvider backend={HTML5Backend}>
          <Provider store={storeState}>
            {children}
          </Provider>
        </DndProvider>
      </ThemeProvider>
    </StyledEngineProvider>
  )

  return render(ui, { wrapper: Wrapper })
}
