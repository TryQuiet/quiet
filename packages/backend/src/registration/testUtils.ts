import { HttpsProxyAgent } from 'https-proxy-agent'
import fetch, { Response } from 'node-fetch'
import PeerId from 'peer-id'
import { createLibp2p } from '../common/testUtils'
import { Storage } from '../storage'

export async function registerUser(csr: string, httpTunnelPort: number, localhost: boolean = true, registrarPort: number = 7789): Promise<Response> {
  let address = '127.0.0.1'
  let options = {
    method: 'POST',
    body: JSON.stringify({ data: csr }),
    headers: { 'Content-Type': 'application/json' }
  }
  if (!localhost) {
    options = Object.assign(options, { agent: new HttpsProxyAgent({ port: httpTunnelPort, host: 'localhost', timeout: 100000 }) })
    address = '4avghtoehep5ebjngfqk5b43jolkiyyedfcvvq4ouzdnughodzoglzad.onion'
    return await fetch(`http://${address}/register`, options)
  }
  return await fetch(`http://${address}:${registrarPort}/register`, options)
}

export const getStorage = async (quietDir: string) => {
  const peerId = await PeerId.create()
  const storage = new Storage(
    quietDir,
    'communityid',
    {
      ...{},
      orbitDbDir: `OrbitDB${peerId.toB58String()}`,
      ipfsDir: `Ipfs${peerId.toB58String()}`
    }
  )
  await storage.init(await createLibp2p(peerId), peerId)
  await storage.initDatabases()
  return storage
}
