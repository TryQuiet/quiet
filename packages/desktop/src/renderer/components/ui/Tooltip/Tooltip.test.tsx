import React from 'react'

import each from 'jest-each'

import { renderComponent } from '../../../testUtils/renderComponent'
import { Tooltip } from './Tooltip'

describe('Tooltip', () => {
  const TooltipContent = () => <div>TooltipContent</div>
  it('renders component', () => {
    const result = renderComponent(
      <Tooltip title='test-title'>
        <TooltipContent />
      </Tooltip>
    )
    expect(result.baseElement).toMatchInlineSnapshot(`
      <body>
        <div>
          <span
            class="css-1fupvor"
          >
            <div>
              TooltipContent
            </div>
          </span>
        </div>
      </body>
    `)
  })

  it('renders component with optional props', () => {
    const result = renderComponent(
      <Tooltip title='test-title' interactive noWrap>
        <TooltipContent />
      </Tooltip>
    )
    expect(result.baseElement).toMatchInlineSnapshot(`
      <body>
        <div>
          <span
            class="css-1fupvor"
          >
            <div>
              TooltipContent
            </div>
          </span>
        </div>
      </body>
    `)
  })

  each(['bottom-start', 'bottom', 'bottom-end', 'top-start', 'top', 'top-end']).test(
    'renders component with correct arrow placement - %s',
    placement => {
      const result = renderComponent(
        <Tooltip title='test-title' placement={placement}>
          <TooltipContent />
        </Tooltip>
      )
      expect(result.baseElement).toMatchSnapshot()
    }
  )
})
