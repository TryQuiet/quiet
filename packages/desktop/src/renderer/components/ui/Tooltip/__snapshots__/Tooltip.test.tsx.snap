// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tooltip renders component with correct arrow placement - bottom 1`] = `
<body>
  <div>
    <span>
      <div>
        TooltipContent
      </div>
    </span>
  </div>
</body>
`;

exports[`Tooltip renders component with correct arrow placement - bottom-end 1`] = `
<body>
  <div>
    <span>
      <div>
        TooltipContent
      </div>
    </span>
  </div>
</body>
`;

exports[`Tooltip renders component with correct arrow placement - bottom-start 1`] = `
<body>
  <div>
    <span>
      <div>
        TooltipContent
      </div>
    </span>
  </div>
</body>
`;

exports[`Tooltip renders component with correct arrow placement - top 1`] = `
<body>
  <div>
    <span>
      <div>
        TooltipContent
      </div>
    </span>
  </div>
</body>
`;

exports[`Tooltip renders component with correct arrow placement - top-end 1`] = `
<body>
  <div>
    <span>
      <div>
        TooltipContent
      </div>
    </span>
  </div>
</body>
`;

exports[`Tooltip renders component with correct arrow placement - top-start 1`] = `
<body>
  <div>
    <span>
      <div>
        TooltipContent
      </div>
    </span>
  </div>
</body>
`;
